### DISCUSSION TITLE
[4일차 - 예진] 호이스팅

### DISCUSSION THREAD

**Discussion by ez1n** [zzangCS/zzangInterview#16] [프론트 면접 질문]
호이스팅이 무엇인가요?

**Comment by smarfy99** [comment]
인터프리터가 변수와 함수의 메모리 공간을 선언 전에 미리 할당하는 것을 의미한다. 변수의 선언과 초기화를 분리한 후에, 선언만 코드만 최상단으로 올리는 것을 뜻한다.

**Comment by ez1n** [comment]
**호이스팅**
자바스크립트에서 선언된 변수, 함수들 모두 끌어올려 유효 범위 내의 최상단에 선언하는 것

**함수 호이스팅**
자바스크립트 변수 생성과 초기화가 분리되어 진행되기 때문에 발생

**Reply by ez1n** [reply]
**호이스팅**

> 변수를 선언하고 초기화했을 때 선언 부분이 최상단으로 끌어올려지는 현상

⇒ 함수 선언식인 경우 식 자체가 통째로 끌어올려진다!!
⇒ 선언과 초기화를 함께 수행하는 경우, 선언 코드까지 실행해야 변수가 초기화된 상태가 된다.

- **원리**

⇒ JS 코드 실행 전 컴파일 과정에서 모든 스코프를 탐색하며 각 스코프의 변수 객체에 여러 식별자를 수집한다.
⇒ 실행 시점으로 넘어가기 전 선언된 식별자에 대한 정보를 이미 알고 있다.

- **규칙**
1. 선언된 함수는 상단에서 참조, 호출이 가능하다.
2. 선언된 var는 상단에서 참조, 할당이 가능하다.
3. 선언된 let, const는 상단에서 참조, 할당이 불가능하다.

- **함수 호이스팅**

⇒ 가장 먼저 이루어짐.
⇒ 함수 선언문은 식별자가 변수 객체에 수집될 때 부가적으로 해당 함수 참조에 대한 초기화까지 자동으로 이루어지기 때문에 상단에서 참조, 호출이 가능하다.

- **변수 호이스팅**

⇒ var : 호이스팅이 발생하면 선언과 초기화가 거의 동시에 이루어진다. (해당 변수에 대한 메모리 존재 → 선언부의 위치에 상관없이 참조, 할당 가능)
⇒ let, const : 호이스팅이 발생하면 선언마 이루어지고 실행 시점에서 선언부를 만날 때까지 초기화가 이루어지지 않는다. (해당 변수에 대한 메모리 존재 X → 선언부 상단에서 참조, 할당 불가)

**Reply by ez1n** [reply]
https://tecoble.techcourse.co.kr/post/2021-04-25-hoisting/

**Comment by skyhanull** [comment]
호이스팅이란 

함수 안에 있는 선언들을 모두 끌어올려서 해당 함수 유효 범위의 최상단에 선언하는 것을 말합니다.
호이스팅은 코드가 실행하기 전 `변수선언/함수선언`이 해당 스코프의 최상단으로 끌어 올려진 것 같은 현상을 말한다.

- `var` 와 `함수 선언문` 이 호이스팅 대상이다.`let` 또는 `const` 그리고 `함수 표현식`은 해당되지 않는다.
- var 나 선언문은 선언과 할당이 동시에 일어나지만 let,const,표현식은 선언과 할당이 따로 일어나서 호이스팅이 되지 않는다
